# Dataquest.io Guided Project One

### In this notebook, data cleaning and exploration techniques are practiced. Specifically, locating and deleting duplicate data are a focus.

#### explore_data() function takes a dataset in the form of a list of lists, and prints out a selection of that dataset given starting and ending points as parameters. Optionally, it can print out the row and column numbers.

In [52]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]
    for row in dataset_slice:
        print(row)
        print('\n')

    if rows_and_columns:
        print('Number of rows: ', len(dataset))
        print('Number of column: ', len(dataset[0]))

#### Import reader from csc. Open and read AppleStore.csv and googleplaystore.csv. Convert those csv files to lists of lists named apple_data and google_data, respectively.

In [53]:
from csv import reader
apple_open = open('AppleStore.csv')
google_open = open('googleplaystore.csv')
apple_read = reader(apple_open)
google_read = reader(google_open)
apple_data = list(apple_read)
google_data = list(google_read)

#### Take a look at the apple_data by calling explore_data().

In [54]:
explore_data(apple_data, 0, len(apple_data[0])-1, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


['429047995', 'Pinterest', '74778624', 'USD', '0.0', '1061

#### Let's take a look at the google_data by calling explore_data().

In [55]:
explore_data(google_data, 0, len(google_data[0]), True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Eve

#### So, from the print out above, there are 13 columns in the google_data set. The  cell below prints out any list in that data set that is not 13 columns wide, in an effort to find missing data. The idea is that within the list of lists, if there is a list that is not the same length as the header, than it is either missing data, or it has additional data.

In [56]:
for row in google_data:
    header_length = len(google_data[0])
    if len(row) != header_length:
        print(row)
        print(google_data.index(row))
        
# Save the header for posterity
google_head = google_data[0]        

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
10473


#### One such row exists. It is deleted below.

In [57]:
del google_data[10473]

#### This cell creates 3 new lists:
#### g_dup_apps contains the names of the apps that have duplicate entries in google_data;
#### g_dup_index contains the index location of the duplicate entry in google_data; 
#### g_unique_apps contains the names of the apps that do not have duplicate entries in google_data

In [58]:
# Create two lists: 1 for unique and 1 for duplicate apps; the list for duplicate apps
# should not have duplicates itself; otherwise, care would have to be taken to make
# sure that 1 row in the dataset with a name in the duplicates list is kept; IE: we do
# not want to remove all the rows that are duplicates; we want to keep 1
g_dup_apps = []
g_unique_apps = []

for row in google_data:
    name = row[0]
    # This if clause keeps duplicates from occurring in the duplicates list
    if name in g_unique_apps and name in g_dup_apps: continue
    elif name not in g_unique_apps: g_unique_apps.append(name)
    elif name in g_unique_apps and name not in g_dup_apps: g_dup_apps.append(name)

print('Number of google apps that have duplicates: ', len(g_dup_apps))
print('\n')
print('Print first 10 duplicates:\n', g_dup_apps[:10])

Number of google apps that have duplicates:  798


Print first 10 duplicates:
 ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Zenefits', 'Google Ads', 'Slack', 'FreshBooks Classic', 'Insightly CRM']


#### The apps with the highest number of downloads are kept in google_data set. All others are deleted.

In [59]:
# remove_dups()
# Function takes in the name of an app w/ duplicates, a dataset, a list, and a
# column number as parameters. It creates a list of the row numbers to be deleted.
# These rows are sorted from least to greatest values of the column number that is
# passed as an argument. The row with the greatest of those values is not deleted. 
def remove_dups(an_app, a_dataset, a_col):

    # Iterate through a_dataset; if the app name matches the one passed as an argument,
    # append its index number to the row, and append the row to dup_list
    dup_list = []
    for i in range(0, len(a_dataset)):
        row = a_dataset[i]
        if row[0] == an_app:
            row.append(i)
            dup_list.append(row)

    # Sort dup_list from greatest to least value in position a_col; delete the row with
    # the greatest value from dup_list
    dup_list.sort(key=lambda x: x[a_col], reverse=True)
    del dup_list[0]

    # Iterate through dup_list; delete each row in a_dataset that corresponds to the index
    # appended to each list in dup_list
    j = 0
    while j < len(dup_list):
        del a_dataset[j]
        j += 1

    return(a_dataset)

In [60]:
# Call remove_dups() on google duplicates
for app in g_dup_apps:
    remove_dups(app, google_data, 3)   

# Check number rows of google_data
print(len(google_data))

9717


In [61]:
for row in apple_data:
    header_length = len(apple_data[0])
    if len(row) != header_length:
        print(row)
        print(apple_data.index(row))
        
# Save the header for posterity
apple_head = apple_data[0]

#### There are no rows that are not the same length as the header in the apple data

In [62]:
a_dup_apps = []
a_unique_apps = []

for row in apple_data:
    name = row[0]
    if name in a_unique_apps:
        if name in a_dup_apps: continue
        else:
            a_dup_apps.append(name)
    else: a_unique_apps.append(name)

print('Number of duplicate apple apps: ', len(a_dup_apps))
print('\n')
print('Print first 10 duplicates:\n', a_dup_apps[:10])

Number of duplicate apple apps:  0


Print first 10 duplicates:
 []


#### There are no duplicate apps in the apple data

In [63]:
# remove_non_eng()
# This function take in a string and returns False if there are non-english characters
# in it; otherwise, it returns True; if the ASCII is greater than 127, then it is
# non-english; no return means that all characters in the string are english
def bool_non_eng(a_string):
    count = 0
    for char in a_string:
        if ord(char) > 127: count += 1
    if count > 3: return False
    else: return True

In [64]:
# Test remove_non_eng()
print(bool_non_eng('Instagram'))                        # True
print(bool_non_eng('爱奇艺PPS -《欢乐颂2》电视剧热播'))      # False
print(bool_non_eng('Docs To Go™ Free Office Suite'))   # True
print(bool_non_eng('Instachat 😜'))                    # True

True
False
True
True


In [65]:
# Call bool_non_eng() on google_data; create list of rows w/ non-english app names;
# append row indices to data to be removed
g_non_eng = []
for row in google_data:
    if bool_non_eng(row[0]) == False:
        row.append(google_data.index(row))
        g_non_eng.append(row)
        
print(len(g_non_eng))
print(g_non_eng[:3])

# 44 non-english app names in google_data

44
[['သိင်္ Astrology - Min Thein Kha BayDin', 'LIFESTYLE', '4.7', '2225', '15M', '100,000+', 'Free', '0', 'Everyone', 'Lifestyle', 'July 26, 2018', '4.2.1', '4.0.3 and up', 469], ['РИА Новости', 'NEWS_AND_MAGAZINES', '4.5', '44274', '8.0M', '1,000,000+', 'Free', '0', 'Everyone', 'News & Magazines', 'August 6, 2018', '4.0.6', '4.4 and up', 2627], ['صور حرف H', 'ART_AND_DESIGN', '4.4', '13', '4.5M', '1,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 27, 2018', '2.0', '4.0.3 and up', 3070]]


In [66]:
# Delete all 44 rows with non-english app names from google_data
for row in g_non_eng:
    index = row[len(row)-1]
    del google_data[index]

# Expect 9,673 rows left
print(len(google_data))

9673


In [67]:
# Call bool_non_eng() on apple_data; create list of rows w/ non-english app names
a_non_eng = []
for row in apple_data:
    if bool_non_eng(row[0]) == False: a_non_eng.append(row)
print(len(a_non_eng))

# No non-english app names in apple_data

0


In [68]:
# Take a look at the cleaned google dataset
explore_data(google_data, 0, 15, True)

# Apple dataset has not changed

['QuickBooks Self-Employed:Mileage Tracker and Taxes', 'FINANCE', '4.4', '15141', '25M', '1,000,000+', 'Free', '0', 'Everyone', 'Finance', 'July 24, 2018', '4.63', '5.0 and up']


['FREE Stock Market Trading Tips', 'FINANCE', '3.9', '714', '3.6M', '50,000+', 'Free', '0', 'Everyone', 'Finance', 'June 7, 2017', '2.10.3', '2.3.3 and up']


['MetaTrader 5', 'FINANCE', '4.5', '42410', 'Varies with device', '1,000,000+', 'Free', '0', 'Everyone', 'Finance', 'August 3, 2018', 'Varies with device', 'Varies with device']


['MetaTrader 4', 'FINANCE', '4.6', '260547', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Finance', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Stock Quote', 'FINANCE', '4.1', '4344', '4.5M', '500,000+', 'Free', '0', 'Everyone', 'Finance', 'July 30, 2018', '3.6.2', '4.0 and up']


['Seeking Alpha', 'FINANCE', '4.5', '22808', 'Varies with device', '1,000,000+', 'Free', '0', 'Everyone', 'Finance', 'August 3, 2018', 'Varies with device', 'Var

#### We only build apps that are free to download and install. So, our main source of revenue is from in-app ads. From here, we need to isolate the free apps from our data sets.

In [69]:
# The element that indicates whether an app is free or not is element #6 in the google_data; and, the value
# of 'Free' indicates this.
g_free = []
for row in google_data:
    if row[6] == 'Free': g_free.append(row)
        
print(g_free[:5])
print("\n")
print(g_free[500:510])
print("\n")
print("Number of free apps in Google Play Store: ", len(g_free))

[['QuickBooks Self-Employed:Mileage Tracker and Taxes', 'FINANCE', '4.4', '15141', '25M', '1,000,000+', 'Free', '0', 'Everyone', 'Finance', 'July 24, 2018', '4.63', '5.0 and up'], ['FREE Stock Market Trading Tips', 'FINANCE', '3.9', '714', '3.6M', '50,000+', 'Free', '0', 'Everyone', 'Finance', 'June 7, 2017', '2.10.3', '2.3.3 and up'], ['MetaTrader 5', 'FINANCE', '4.5', '42410', 'Varies with device', '1,000,000+', 'Free', '0', 'Everyone', 'Finance', 'August 3, 2018', 'Varies with device', 'Varies with device'], ['MetaTrader 4', 'FINANCE', '4.6', '260547', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Finance', 'July 31, 2018', 'Varies with device', 'Varies with device'], ['Stock Quote', 'FINANCE', '4.1', '4344', '4.5M', '500,000+', 'Free', '0', 'Everyone', 'Finance', 'July 30, 2018', '3.6.2', '4.0 and up']]


[['Chictopia', 'LIFESTYLE', '4.1', '360', '4.6M', '10,000+', 'Free', '0', 'Everyone', 'Lifestyle', 'April 30, 2014', '1.6.2', '2.2 and up', 1287], ['Scarf Fashion 

In [70]:
# The element that indicates whether an app is free or not is element #4 in the appgle_data; and, the value
# of '0.0' indicates this.
a_free = []
for row in apple_data:
    if row[4] == '0.0': a_free.append(row)

print(a_free[:5])
print(len(a_free))

[['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1'], ['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1'], ['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1'], ['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1'], ['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']]
4056


### Analysis

#### Validation Strategy:
#### 1. Build a minimal Android version of the app, and add it to Google Play.
#### 2. If the app has a good response from users, we develop it further.
#### 3. If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

#### Because our apps generate revenue from the number of people that are using them, we are looking for the types of apps that are best able to attract the most users. Because our goal is release an app in both markets, we are looking for app profiles that are successful on both platforms. We begin by looking at frequency tables for app genres from each store to get an idea of the most common genres in each market.

In [71]:
print(google_head)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


In [72]:
print(apple_head)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


#### Genre Frequency Tables

In [91]:
# freq_table()
# Function takes a dataset, an index to count data in, and it has a second index parameter set as
# by default to None; if b_index == None, it returns a list of data pertaining to the values in
# the a_index; this is a dictionary of frequencies followed by figure of the total count to be
# used in percentage calculations; If b_index != None, it returns a list containing a_table,
# a_total, b_table, b_total, ab_table, and ab_total
def freq_table(a_dataset, a_index, b_index=None):
    
    a_table = {}
    b_table = {}
    ab_table = {}

    # Count the occurrences
    for row in a_dataset:
        if row[a_index] in a_table: a_table[row[a_index]] += 1
        else: a_table[row[a_index]] = 1
        if b_index != None:    
            if row[b_index] in b_table: b_table[row[b_index]] += 1
            else: b_table[row[b_index]] = 1
            ab_combined = row[a_index] + ", " + row[b_index]
            if ab_combined in ab_table: ab_table[ab_combined] += 1
            else: ab_table[ab_combined] = 1
    
    # To store the sume of all occurrences
    a_sum = 0
    b_sum = 0
    ab_sum = 0
    
    # Sum values in a_table
    for key in a_table:
        a_sum += a_table[key]
    
    # If argument given for b_index, sum b and ab tables
    if b_index != None:
        for key in b_table:
            b_sum += b_table[key]
        for key in ab_table:
            ab_sum += ab_table[key]
    
    # Change values in a_table to frequencies
    for key in a_table:
        a_table[key] = (a_table[key] / a_sum) * 100
    
    # If value passed for b_index, change values in b and ab tables to freq's
    if b_index != None:
        for key in b_table:
            b_table[key] = (b_table[key] / b_sum) * 100
        for key in ab_table:
            ab_table[key] = (ab_table[key] / ab_sum) * 100
            
    # If b_index == None, return list of a data
    if b_index == None:
        a_list = [a_table]
        return a_list
    else:
        # If b_index != None, return all a, b, and ab data
        ab_list = [a_table, b_table, ab_table]
        return ab_list


In [136]:
def print_table(a_dataset, a_index, b_index=None):

    # Call freq_table() function w/ values passed; save output
    list_tables = freq_table(a_dataset, a_index, b_index)

    # Separate a_list just in case there are others
    a_list = list_tables[0]

    # Create list of tuples from dictionary
    a_freqs = []
    for key in a_list:
        tup = (a_list[key], key)
        a_freqs.append(tup)

    # Sort a_freqs
    a_freqs = sorted(a_freqs, reverse=True)

    # Print items in sorted a_list
    print("a_index Frequencies:")
    print(20 * "-", "\n")
    for item in a_freqs:
        print(item[1], " : ", item[0])
    print("\n")

    # Check to see if b_index value was passed; sort; print
    if len(list_tables) > 1:

        # Create lists to print freq's
        b_freqs = []
        ab_freqs = []

        # Append dictionary keys, vals to b_list
        b_list = list_tables[1]
        for key in b_list:
            tup = (b_list[key], key)
            b_freqs.append(tup)

        # Sort b_freqs
        b_freqs = sorted(b_freqs, reverse=True)

        # Print items in b_list
        print("b_index Frequencies:")
        print(20 * "-", "\n")
        for item in b_freqs:
            print(item[1], " : ", item[0])
        print("\n")

        # Append dictionary keys, vals to ab_list
        ab_list = list_tables[2]
        for key in ab_list:
            tup = (ab_list[key], key)
            ab_freqs.append(tup)

        # Sort ab_freqs
        ab_freqs = sorted(ab_freqs, reverse=True)

        # Print items in ab_list
        print("ab_index Frequencies:")
        print(21 * "-", "\n")
        for item in ab_freqs:
            print(item[1], " : ", item[0])

In [144]:
# Create frequency table for the genre for iOS apps
# Element #11 contains the prime genre; #5 contains the number of ratings
print("Apple App Store")
print("\n")
print_table(a_free, 11)

Apple App Store


a_index Frequencies:
-------------------- 

Games  :  55.64595660749507
Entertainment  :  8.234714003944774
Photo & Video  :  4.117357001972387
Social Networking  :  3.5256410256410255
Education  :  3.2544378698224854
Shopping  :  2.983234714003945
Utilities  :  2.687376725838264
Lifestyle  :  2.3175542406311638
Finance  :  2.0710059171597637
Sports  :  1.947731755424063
Health & Fitness  :  1.8737672583826428
Music  :  1.6518737672583828
Book  :  1.6272189349112427
Productivity  :  1.5285996055226825
News  :  1.4299802761341223
Travel  :  1.3806706114398422
Food & Drink  :  1.0601577909270217
Weather  :  0.7642998027613412
Reference  :  0.4930966469428008
Navigation  :  0.4930966469428008
Business  :  0.4930966469428008
Catalogs  :  0.22189349112426035
Medical  :  0.19723865877712032




#### Among the free, English apps, 55.65 percent are games. The genre with the next highest frequency in the App Store is Entertainment with 8.24 percent. Third and fourth are Photo & Video and Social Networking with 4.12 and 3.53 percent, respectively. So, the overwhelming majority of apps are games. Clearly, the most popular apps are those used for entertainment.

#### Among the least frequent apps in the App Store are those that are used for productivity. These include navigation, medical, news and books. The three least frequent apps are for business, catalogs, and medical. And, they comprise 0.49, 0.22

In [141]:
# Create frequency table for the number of installations per genre for Android apps
# Element #5 contains the number of installations
# Element #1 contains the Category which seems to be the same info contained in element #6
print("Google Play Store")
print("\n")
print_table(g_free, 1, 5)

Google Play Store


a_index Frequencies:
-------------------- 

FAMILY  :  19.88527724665392
GAME  :  11.910921156225395
TOOLS  :  8.581711843437184
PRODUCTIVITY  :  4.4314475312113375
LIFESTYLE  :  4.060285682150489
SPORTS  :  4.03779102463165
MEDICAL  :  3.9590597233157125
BUSINESS  :  3.767855134405579
HEALTH_AND_FITNESS  :  3.655381846811382
PHOTOGRAPHY  :  3.5091665729389274
PERSONALIZATION  :  3.4529299291418285
SOCIAL  :  3.2842199977505344
NEWS_AND_MAGAZINES  :  3.11551006635924
FINANCE  :  3.070520751321561
SHOPPING  :  2.9018108199302666
TRAVEL_AND_LOCAL  :  2.766842874817231
COMMUNICATION  :  2.4631649983129007
VIDEO_PLAYERS  :  1.9120458891013385
BOOKS_AND_REFERENCE  :  1.7208413001912046
MAPS_AND_NAVIGATION  :  1.4846473962433921
FOOD_AND_DRINK  :  1.4059160949274547
HOUSE_AND_HOME  :  0.9897649308289281
LIBRARIES_AND_DEMO  :  0.9447756157912496
WEATHER  :  0.8323023281970533
PARENTING  :  0.652345068046339
AUTO_AND_VEHICLES  :  0.37116184906084804
COMICS  :  0.23619390394

#### Games dominate the Play Store, as well at 55.65 percent of the market. Entertainment and Photo & Video apps are the next largest segments at 8.24 and 4.12 percent, respectively.

#### We can see that the largest portion of apps in the Play Store are those which have been installed at least 1M times but less than 10M. These apps comprise 14.34 percent of the market. The next two largest installation segments are 10M+ and 100K+ which comprise 12.10 and 10.22 percent, respectively.

#### For the Google Play Store, we have the number of installations included in the available data. Although it is imperfect in that it bins the values very broadly. For example, the bin that contains the most installations is 10M+ and the next smallest bin is 1M+. And, the third largest bin is 100K+. So, this data are discretized, and the bins not as narrow as they otherwise could be. However, it is still useful to look at app genres as they relate to this installation information. For example, we can see below that 3.01 percent of all of the apps in the Play Store are family apps that have been installed at least 1M times, but less than 10M times. This is the largest segment of these combined categories. Family apps that have been installed 100K+ times but less than 1M times comprise 2.56 percent of the Play Store. And, games that have been installed 10M+ times are 2.51 percent of the market. It is worth noting that of the 11 largest segments of genre and installation categories combined, 7 are family apps and 4 are game apps. And, they account for more than 20 percent of the apps in the Play Store.

In [180]:
# Pull dictionary from frequency table
genres_ios = freq_table(a_free, 11)[0]

# Loop through the genres in the dictionary
for genre in genres_ios:
    total = 0
    len_genre = 0
    # Loop through rows of a_free; if the genre == that from dictionary...
    for app in a_free:
        if app[-5] == genre:
            # ... Increment total and length figures
            total += float(app[5])
            len_genre += 1
    avg_number = total / len_genre
    print(genre, ":", avg_number)

Social Networking : 53078.195804195806
Photo & Video : 27249.892215568863
Games : 18924.68896765618
Music : 56482.02985074627
Reference : 67447.9
Health & Fitness : 19952.315789473683
Weather : 47220.93548387097
Utilities : 14010.100917431193
Travel : 20216.01785714286
Shopping : 18746.677685950413
News : 15892.724137931034
Navigation : 25972.05
Lifestyle : 8978.308510638299
Entertainment : 10822.961077844311
Food & Drink : 20179.093023255813
Sports : 20128.974683544304
Book : 8498.333333333334
Finance : 13522.261904761905
Education : 6266.333333333333
Productivity : 19053.887096774193
Business : 6367.8
Catalogs : 1779.5555555555557
Medical : 459.75


#### Using the number of ratings as a proxy for the number of installations, we can see that Reference, Music, and Social Networking are the three most installed apps in the Apple Apps Store with Reference being the most installed and Social Networking being the least installed of those three.

#### Let's dig a bit further by looking at the apps that have the most installations within each of these three categories.

In [183]:
# Loop through the dataset; print the apps w/in the Navigation category
for app in a_free:
    if app[-5] == 'Reference':
        print(app[1], ':', app[5])

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
彩库宝典-【官方版】 : 0
Jishokun-Japanese English Dictionary & Translator : 0
無料で音楽や写真・カメラの裏技アプリ for iPhone7 : 0


#### It looks like Christian and Islamic references are very popular. Among the more secular apps are Dictionaries and Google Translate.

#### There is also, a popular constellation reference app, a Pokemon app, and several Minecraft reference apps.

In [185]:
# Loop through the dataset; print the apps w/in the Music category
for app in a_free:
    if app[-5] == 'Music':
        print(app[1], ':', app[5])

Pandora - Music & Radio : 1126879
Spotify Music : 878563
Shazam - Discover music, artists, videos & lyrics : 402925
iHeartRadio – Free Music & Radio Stations : 293228
SoundCloud - Music & Audio : 135744
Magic Piano by Smule : 131695
Smule Sing! : 119316
TuneIn Radio - MLB NBA Audiobooks Podcasts Music : 110420
Amazon Music : 106235
SoundHound Song Search & Music Player : 82602
Sonos Controller : 48905
Bandsintown Concerts : 30845
Karaoke - Sing Karaoke, Unlimited Songs! : 28606
My Mixtapez Music : 26286
Sing Karaoke Songs Unlimited with StarMaker : 26227
Ringtones for iPhone & Ringtone Maker : 25403
Musi - Unlimited Music For YouTube : 25193
AutoRap by Smule : 18202
Spinrilla - Mixtapes For Free : 15053
Napster - Top Music & Radio : 14268
edjing Mix:DJ turntable to remix and scratch music : 13580
Free Music - MP3 Streamer & Playlist Manager Pro : 13443
Free Piano app by Yokee : 13016
Google Play Music : 10118
Certified Mixtapes - Hip Hop Albums & Mixtapes : 9975
TIDAL : 7398
YouTube Mu

#### Not surprisingly, the top five most installed music apps are:
- Pandora - Music & Radio : 1126879
- Spotify Music : 878563
- Shazam - Discover music, artists, videos & lyrics : 402925
- iHeartRadio – Free Music & Radio Stations : 293228
- SoundCloud - Music & Audio : 135744

In [186]:
# Loop through the dataset; print the apps w/in the Social Networking category
for app in a_free:
    if app[-5] == 'Social Networking':
        print(app[1], ':', app[5])

Facebook : 2974676
Pinterest : 1061624
Skype for iPhone : 373519
Messenger : 351466
Tumblr : 334293
WhatsApp Messenger : 287589
Kik : 260965
ooVoo – Free Video Call, Text and Voice : 177501
TextNow - Unlimited Text + Calls : 164963
Viber Messenger – Text & Call : 164249
Followers - Social Analytics For Instagram : 112778
MeetMe - Chat and Meet New People : 97072
We Heart It - Fashion, wallpapers, quotes, tattoos : 90414
InsTrack for Instagram - Analytics Plus More : 85535
Tango - Free Video Call, Voice and Chat : 75412
LinkedIn : 71856
Match™ - #1 Dating App. : 60659
Skype for iPad : 60163
POF - Best Dating App for Conversations : 52642
Timehop : 49510
Find My Family, Friends & iPhone - Life360 Locator : 43877
Whisper - Share, Express, Meet : 39819
Hangouts : 36404
LINE PLAY - Your Avatar World : 34677
WeChat : 34584
Badoo - Meet New People, Chat, Socialize. : 34428
Followers + for Instagram - Follower Analytics : 28633
GroupMe : 28260
Marco Polo Video Walkie Talkie : 27662
Miitomo : 2

#### As for the Social Networking apps, there is little surprise that the top six most installed are the following:
- Facebook : 2974676
- Pinterest : 1061624
- Skype for iPhone : 373519
- Messenger : 351466
- Tumblr : 334293
- WhatsApp Messenger : 287589

In [189]:
google_head

['App',
 'Category',
 'Rating',
 'Reviews',
 'Size',
 'Installs',
 'Type',
 'Price',
 'Content Rating',
 'Genres',
 'Last Updated',
 'Current Ver',
 'Android Ver']

In [199]:
# Pull dictionary from frequency table
genres_android = freq_table(g_free, 1)[0]

# Loop through the genres in the dictionary
for genre in genres_android:
    total = 0
    len_genre = 0
    # Loop through rows of a_free; if the genre == that from dictionary...
    for app in g_free:
        if app[1] == genre:
            # ... Increment total and length figures
            total += float(app[5].replace('+', '').replace(',', ''))
            #print(total)
            len_genre += 1
    avg_number = total / len_genre
    print(genre, ":", avg_number)

FINANCE : 1539571.912087912
FOOD_AND_DRINK : 2190710.008
HEALTH_AND_FITNESS : 4869225.852307692
HOUSE_AND_HOME : 1917187.0568181819
LIBRARIES_AND_DEMO : 749950.119047619
LIFESTYLE : 1485774.0138504156
GAME : 33100967.374881964
FAMILY : 5769044.48020362
MEDICAL : 148400.01136363635
SOCIAL : 48184458.56849315
SHOPPING : 12588522.03488372
PHOTOGRAPHY : 32321374.407051284
SPORTS : 4874455.941504179
TRAVEL_AND_LOCAL : 27921561.32520325
TOOLS : 15007920.54259502
PERSONALIZATION : 7557723.087947883
PRODUCTIVITY : 35973802.82741117
PARENTING : 542603.6206896552
WEATHER : 5747142.162162162
VIDEO_PLAYERS : 36599595.4117647
NEWS_AND_MAGAZINES : 27058525.84837545
MAPS_AND_NAVIGATION : 5486066.590909091
COMMUNICATION : 46932448.40639269
BOOKS_AND_REFERENCE : 3829909.54248366
ART_AND_DESIGN : 794392.3076923077
AUTO_AND_VEHICLES : 310910.9393939394
BUSINESS : 739077.5223880596
DATING : 4022500.0
COMICS : 935007.1428571428
EVENTS : 1381.111111111111
BEAUTY : 426420.8333333333


#### Above we can see that the four most installed categories are:
- SOCIAL : 48184458.56849315
- COMMUNICATION : 46932448.40639269
- PRODUCTIVITY : 35973802.82741117
- GAME : 33100967.374881964

#### Now lets take a look at the most installed apps within each of these four categories.

In [210]:
# Loop through the dataset; print the apps w/in the Social Networking category
for app in g_free:
    if app[1] == 'SOCIAL':
        print(app[0], ':', app[5].replace('+', ''))

Facebook : 1,000,000,000
Instagram : 1,000,000,000
Facebook Lite : 500,000,000
Messages, Text and Video Chat for Messenger : 10,000,000
Tumblr : 100,000,000
All Social Networks : 1,000,000
Snapchat : 500,000,000
Social network all in one 2018 : 100,000
Pinterest : 100,000,000
TextNow - free text + calls : 10,000,000
Google+ : 1,000,000,000
The Messenger App : 1,000,000
Messenger Pro : 1,000,000
Free Messages, Video, Chat,Text for Messenger Plus : 1,000,000
Telegram X : 5,000,000
The Video Messenger App : 100,000
Jodel - The Hyperlocal App : 1,000,000
Who Viewed My Facebook Profile - Stalkers Visitors : 5,000,000
Hide Something - Photo, Video : 5,000,000
Love Sticker : 1,000,000
Web Browser & Fast Explorer : 5,000,000
LiveMe - Video chat, new friends, and make money : 10,000,000
VidStatus app - Status Videos & Status Downloader : 5,000,000
Love Images : 1,000,000
Web Browser ( Fast & Secure Web Explorer) : 500,000
SPARK - Live random video chat & meet new people : 5,000,000
Golden teleg

In [219]:
# Loop through the dataset; print the apps w/in the Social Networking category
for app in g_free:
    if app[1] == 'COMMUNICATION':
        print(app[0], ':', app[5])

Azar : 50,000,000+
LokLok: Draw on a Lock Screen : 500,000+
WhatsApp Messenger : 1,000,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Discord - Chat for Gamers : 10,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Firefox Browser fast & private : 100,000,000+
Microsoft Edge : 5,000,000+
Gmail : 1,000,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
imo beta free calls and text : 100,000,000+
Hangouts : 1,000,000,000+
AntennaPict β : 1,000,000+
imo free video calls and chat : 500,000,000+
free video calls and chat : 50,000,000+
Skype - free IM & video calls : 1,000,000,000+
Talkatone: Free Texts, Calls & Phone Number : 10,000,000+
Kik : 100,000,000+
K-@ Mail - Email App : 100,000+
KakaoTalk: Free Calls & Text : 100,000,000+
K-9 Material (unofficial) : 5,000+
M star Dialer : 100,000+
Free WiFi Con

In [220]:
# Loop through the dataset; print the apps w/in the Social Networking category
for app in g_free:
    if app[1] == 'PRODUCTIVITY':
        print(app[0], ':', app[5])

Microsoft Word : 500,000,000+
All-In-One Toolbox: Cleaner, Booster, App Manager : 10,000,000+
Adobe Acrobat Reader : 100,000,000+
AVG Cleaner – Speed, Battery & Memory Booster : 10,000,000+
Google Drive : 1,000,000,000+
QR Scanner & Barcode Scanner 2018 : 10,000,000+
Chrome Beta : 10,000,000+
Microsoft Outlook : 100,000,000+
Google PDF Viewer : 10,000,000+
Microsoft Excel : 100,000,000+
My Claro Peru : 5,000,000+
Power Booster - Junk Cleaner & CPU Cooler & Boost : 1,000,000+
Google Assistant : 10,000,000+
Microsoft OneDrive : 100,000,000+
Calculator - unit converter : 50,000,000+
Microsoft OneNote : 100,000,000+
Metro name iD : 10,000,000+
Google Keep : 100,000,000+
Archos File Manager : 5,000,000+
ES File Explorer File Manager : 100,000,000+
ASUS SuperNote : 10,000,000+
HTC File Manager : 10,000,000+
MyMTN : 1,000,000+
Dropbox : 500,000,000+
ASUS Quick Memo : 10,000,000+
HTC Calendar : 10,000,000+
Google Calendar : 500,000,000+
Google Docs : 100,000,000+
ASUS Calling Screen : 10,000,0

In [221]:
# Loop through the dataset; print the apps w/in the Social Networking category
for app in g_free:
    if app[1] == 'GAME':
        print(app[0], ':', app[5])

ROBLOX : 100,000,000+
Subway Surfers : 1,000,000,000+
Candy Crush Saga : 500,000,000+
Solitaire : 10,000,000+
Bubble Shooter : 10,000,000+
Hello Kitty Nail Salon : 50,000,000+
slither.io : 100,000,000+
Clash Royale : 100,000,000+
Temple Run 2 : 500,000,000+
Pou : 500,000,000+
Helix Jump : 100,000,000+
Block Puzzle : 5,000,000+
Angry Birds Rio : 100,000,000+
Plants vs. Zombies FREE : 100,000,000+
Sonic Dash : 100,000,000+
Candy Crush Soda Saga : 100,000,000+
Zombie Hunter King : 1,000,000+
Clash of Clans : 100,000,000+
Kick the Buddy : 50,000,000+
Block Puzzle Classic Legend ! : 5,000,000+
PAC-MAN : 100,000,000+
Super Jim Jump - pixel 3d : 1,000,000+
8 Ball Pool : 100,000,000+
Magic Tiles 3 : 50,000,000+
Bubble Witch 3 Saga : 50,000,000+
Word Search : 10,000,000+
Granny : 50,000,000+
Angry Birds Classic : 100,000,000+
Flow Free : 100,000,000+
Race the Traffic Moto : 10,000,000+
Fishdom : 10,000,000+
Galaxy Attack: Alien Shooter : 10,000,000+
Zombie Tsunami : 100,000,000+
Bubble Shooter 